optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -67,16 +68,44 @@ public boolean isConstraintMatchEnabled() {
     // Worker methods
     // ************************************************************************
 
+    protected class MultiLevelConstraintUndoListener implements ActivationUnMatchListener {
+
+        private Map<Integer, ActivationUnMatchListener> activationUnMatchListenerMap;
+
+        public MultiLevelConstraintUndoListener(int scoreLevel, ActivationUnMatchListener activationUnMatchListener) {
+            this.activationUnMatchListenerMap = new HashMap<Integer, ActivationUnMatchListener>();
+            match(scoreLevel, activationUnMatchListener);
+        }
+
+        @Override
+        public final void unMatch(RuleRuntime ruleRuntime, Match match) {
+            for (Integer scoreLevel : activationUnMatchListenerMap.keySet()) {
+                unMatch(scoreLevel);
+            }
+        }
+
+        public void unMatch(int scoreLevel) {
+            if (activationUnMatchListenerMap.containsKey(scoreLevel)) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                activationUnMatchListenerMap.get(scoreLevel).unMatch(null, null);
+                activationUnMatchListenerMap.remove(scoreLevel);
```

Comment:
containsKey, get and remove each do a lookup of the scoreLevel, so the lookup happens 3 times per scoreLevel. Replacing with entrySet() and clear() usage.